xdsl/rewriting/composable_rewriting/immutable_ir/immutable_ir.py

```diff
@@ -0,0 +1,550 @@
+        # to the correct block
+        for block in blocks:
+            block_map[block] = IBlock([],[])
+
+        immutable_blocks = [
+            IBlock.from_mutable(block, value_map, block_map)
+            for block in blocks
+        ]
+
+        assert (blocks[0].parent is not None)
+        region = IRegion(immutable_blocks)
+
+        # clean up successor references to blocks for ops inside this region
+        for block, imm_block in zip(blocks, region.blocks):
+            dummy_block = block_map[block]
+            for op in region.ops:
+                try:
+                    dummy_index = op.successors.index(dummy_block)
+                except ValueError:
+                    continue
+                # replace dummy successor with actual successor
+                object.__setattr__(op, "successors", IList(op.successors[:dummy_index]+[imm_block]+op.successors[dummy_index+1:]))
+
+        return region
+
+    def get_mutable_copy(
+            self,
+            value_mapping: Optional[dict[ISSAValue, SSAValue]] = None,
+            block_mapping: Optional[dict[IBlock, Block]] = None) -> Region:
+        """
+        Returns a mutable region that is a copy of this immutable region.
+        The value_mapping and block_mapping are used to map already known correspondings
+        of immutable values to mutable values and immutable blocks to mutable blocks.
+        """
+        if value_mapping is None:
+            value_mapping = {}
+        if block_mapping is None:
+            block_mapping = {}
+        mutable_blocks: List[Block] = []
+        # All mutable blocks have to be initialized first so that ops can
+        # refer to them in their successor lists.
+        for block in self.blocks:
+            mutable_blocks.append(mutable_block := Block.from_arg_types(block.arg_types))
+            block_mapping[block] = mutable_block
+        for block in self.blocks:
+            # This will use the already created Block and populate it
+            block.get_mutable_copy(value_mapping=value_mapping, block_mapping=block_mapping)
+        return Region.from_block_list(mutable_blocks)
+
+
+@dataclass(frozen=True)
+class IBlock:
+    """An immutable block contains a list of immutable operations. IBlocks are contained in IRegions."""
+    args: IList[IBlockArg]
+    ops: IList[IOp]
+
+    @property
+    def arg_types(self) -> List[Attribute]:
+        frozen_arg_types = [arg.typ for arg in self.args]
+        return frozen_arg_types
+
+    def __hash__(self) -> int:
+        return (id(self))
+
+    def __eq__(self, __o: object) -> bool:
+        return self is __o
+
+    def __repr__(self) -> str:
+        return "block of" + str(len(self.ops)) + " with args: " + str(
+            self.args)
+
+    def __post_init__(self):
+        for arg in self.args:
+            object.__setattr__(arg, "block", self)
+
+    def __init__(self, args: Sequence[Attribute] | Sequence[IBlockArg],
```

Comment:
What is the case where we give a list of `IBlockArg`?

Reply:
This is intended for during rewriting in the case that existing IBlockArgs should be reused. I am not quite 100% sure reuse still works in this design, as the IBlockArg holds a reference to its defining IBlock.
I am not quite sure whether some rewrites are expressed easier when the Blockargs can be created independently from the block first. So for now I am removing this.